cms/djangoapps/contentstore/tasks.py

```diff
@@ -1117,101 +1120,155 @@ def validate_user():
-                return "ACCESS_DENIED"
-            else:
-                return "FAILURE"
-        except requests.exceptions.RequestException as e:
-            return "FAILURE"
-
-    def verify_url(url):
-        """Returns true if url request returns 200"""
-        try:
-            response = requests.get(url, timeout=5)
-            return response.status_code == 200
-        except requests.exceptions.RequestException as e:
-            return False
-
-    def scan_course(course_key):
+    def scan_course_for_links(course_key):
         """
-        Scans the course and returns broken link tuples.
-          [<block_id>, <broken_link>]
+        Returns a list of all urls in a course.
+        Returns: [ [block_id1, url1], [block_id2, url2], ... ]
         """
-        broken_links = []
         verticals = modulestore().get_items(course_key, qualifiers={'category': 'vertical'}, revision=ModuleStoreEnum.RevisionOption.published_only)
-        print('verticals: ', verticals)
         blocks = []
+        urls_to_validate = []
 
         for vertical in verticals:
             blocks.extend(vertical.get_children())
 
         for block in blocks:
-            usage_key = block.usage_key
+            block_id = str(block.usage_key)
             block_info = get_block_info(block)
             block_data = block_info['data']
-            urls = get_urls(block_data)
 
-            for url in urls:
-                if url == '#':
-                    break
-                standardized_url = convert_to_standard_url(url, course_key)
-                # TODO if check_url ACCESS_DENIED and is studio link, it's locked
-                if not verify_url(standardized_url):
-                    broken_links.append([str(usage_key), url])
+            url_list = get_urls(block_data)
+            urls_to_validate += [[block_id, url] for url in url_list]
 
-        return broken_links
+        return urls_to_validate
+
+    async def validate_url_access(session, url_data, course_key):
+        """
+        Returns the status of a url request
+        Returns: {block_id1, url1, status}
+        """
+        block_id, url = url_data
+        result = {'block_id': block_id, 'url': url}
+        standardized_url = convert_to_standard_url(url, course_key)
+        try:
+            async with session.get(standardized_url, timeout=5) as response:
+                result.update({'status': response.status})
+        except Exception as e:
+            result.update({'status': None})
+            LOGGER.debug(f'[Link Check] Request error when validating {url}: {str(e)}')
+        return result
+
+    async def validate_urls_access_in_batches(url_list, course_key, batch_size=100):
+        """
+        Returns the statuses of a list of url requests.
+        Returns: [ {block_id1, url1, status}, {block_id2, url2, status}, ... ]
+        """
+        responses = []
+        url_count = len(url_list)
+
+        for i in range(0, url_count, batch_size):
```

Comment:
Tell me if I read this correctly, there might be a bug.
This is a loop that goes from 0 to batch size, which is 100.
For each iteration, a batch starts at index i and ends at index i + 100.
And then you go over all of the urls.

That would mean:
Iteration 1 checks URLs 0 to 99
Iteration 2 checks URLs 1 to 100
Iteration 3 checks URLs 2 to 101
... and so on

So it looks to me like instead of checking urls 0-99, then 100-199, it re-checks the same URLs over and over and then stops at 199 without finishing if there are more URLs. Did I get that right?

Reply:
It should be a loop that goes from 0 to url_count, with a step of batch_size (which is 100).
Iteration 1: i = 0
Iteration 2: i = 100
Iteration 3: i = 200

Reply:
You're right, I missed the last argument in the range function.